codegen/post_gen_hooks.go

```diff
@@ -23,114 +23,229 @@ package codegen
 import (
 	"fmt"
 	"path"
+	"path/filepath"
+	"sync"
 
+	"encoding/json"
 	"github.com/pkg/errors"
-	"path/filepath"
 )
 
+const clientInterface = "Client"
+
+// helper struct to pull out the fixture config
+type moduleConfig struct {
+	Config *config `json:"config"`
+}
+
+// config is the struct corresponding to the config field in client-config.json
+type config struct {
+	CustomImportPath string   `json:"customImportPath"`
+	Fixture          *Fixture `json:"fixture"`
+}
+
+// Fixture specifies client fixture import path and all scenarios
+type Fixture struct {
+	// ImportPath is the package where the user-defined Fixture global variable is contained.
+	// The Fixture object defines, for a given client, the standardized list of fixture scenarios for that client
+	ImportPath string `json:"importPath"`
+	// Scenarios is a map from zanzibar's exposed method name to a list of user-defined fixture scenarios for a client
+	Scenarios map[string][]string `json:"scenarios"`
+}
+
+// Validate the fixture configuration
+func (f *Fixture) Validate(methods map[string]interface{}) error {
+	if f.ImportPath == "" {
+		return errors.New("fixture importPath is empty")
+	}
+	for m := range f.Scenarios {
+		if _, ok := methods[m]; !ok {
+			return errors.Errorf("%q is not a valid method", m)
+		}
+	}
+	return nil
+}
+
 // ClientMockGenHook returns a PostGenHook to generate client mocks
-func ClientMockGenHook(h *PackageHelper) (PostGenHook, error) {
-	bin, err := NewMockgenBin(h.PackageRoot())
+func ClientMockGenHook(h *PackageHelper, t *Template) (PostGenHook, error) {
+	bin, err := NewMockgenBin(h, t)
 	if err != nil {
 		return nil, errors.Wrap(err, "error building mockgen binary")
 	}
 
 	return func(instances map[string][]*ModuleInstance) error {
 		fmt.Println("Generating client mocks:")
 		mockCount := len(instances["client"])
+		ec := make(chan error, mockCount)
+		var files sync.Map
+		var wg sync.WaitGroup
+		wg.Add(mockCount)
 		for i, instance := range instances["client"] {
-			if err := bin.GenMock(
-				instance,
-				"mock-client/mock_client.go",
-				"clientmock",
-				"Client",
-			); err != nil {
-				return errors.Wrapf(
-					err,
-					"error generating mocks for client %q",
+			go func(instance *ModuleInstance, i int) {
+				defer wg.Done()
+
+				var mc moduleConfig
+				if err := json.Unmarshal(instance.JSONFileRaw, &mc); err != nil {
+					ec <- errors.Wrapf(
+						err,
+						"error parsing client-config.json for client %q",
+						instance.InstanceName,
+					)
+					return
+				}
+
+				buildDir := h.CodeGenTargetPath()
+				genDir := filepath.Join(buildDir, instance.Directory, "mock-client")
+
+				importPath := instance.PackageInfo.GeneratedPackagePath
+				if instance.ClassType == "custom" {
+					importPath = mc.Config.CustomImportPath
+					if importPath == "" {
```

Comment:
hmm, the logic here is to decide where the interface is defined: custom import path if type is custom otherwise generated import path. Not sure how validators can avoid the check.

Reply:
Yes, it is already validated here https://github.com/uber/zanzibar/blob/master/codegen/gateway.go#L260, I will just delete this check.

Reply:
I meant something like https://godoc.org/gopkg.in/validator.v2?